test: Extend test coverage of BIP125 and document confusing/inconsistent behavior

mjdietzx opened the pull request:
Intention of this PR is to extend test coverage of BIP125. Mostly around [Rule 5](https://github.com/bitcoin/bips/blob/master/bip-0125.mediawiki#implementation-details). Focus is on highlighting confusing/inconsistent behavior that I think we need to consider as we move forward with multiple PRs (and different approaches to BIP125 in general) that are under review.

This PR also splits tests out of #22698 (which I've made a draft for now until some related PRs are merged) as they are generally useful and highlight some behavior I think is worth considering.

All that said, I do not see any reason we can't eliminate all these inconsistencies while keeping inherited signaling, which is what #22698 does. But either way, making it clear in tests at least motivates a longer-term fix

mjdietzx committed:
test: clarify no inherited replaceability assertions

mjdietzx committed:
test: replacement tx rejected bc conflicts may be double counted

mjdietzx committed:
test: dynamic rbf inherited signaling in the case of a reorg

In the event of a reorg, the assumption that a newly added tx has no
in-mempool children is false. If the children opted-out of rbf they
would show \`RBFTransactionState::FINAL\` prior to the reorg.
Upon their rbf opt-in parent being accepted back into the mempool, they
would show \`RBFTransactionState::REPLACEABLE_BIP125\`.

laanwj commented:
Concept ACK, thanks for adding tests

fanquake commented:
cc @ariard @darosior

mjdietzx commented:
Rebased

pg156 commented:
Concept ACK. Tested on macOS Big Sur.

pg156 commented on test/functional/feature_rbf.py:
```diff
@@ -638,5 +656,129 @@ def test_replacement_relay_fee(self):
+        # replaceability, along with _all_ of its descendants.
+        entry = self.nodes[0].getmempoolentry(parent_txs[0]["txid"])
+        assert_greater_than(entry['descendantcount'], MAX_REPLACEMENT_LIMIT)
+        assert_equal(True, self.nodes[0].getmempoolentry(parent_txs[0]["txid"])['bip125-replaceable'])
```
Minor suggestion, otherwise it is possibly unclear if 0 is opt-in and 1 is opt-out, or vice versa.
```suggestion
        assert_equal(True, self.nodes[0].getmempoolentry(optin_parent_tx["txid"])['bip125-replaceable'])
```

pg156 commented on test/functional/feature_rbf.py:
```diff
@@ -127,6 +133,20 @@ def make_utxo(self, node, amount, confirmed=True, scriptPubKey=DUMMY_P2WPKH_SCRI
 
+    def create_multiple_input_self_transfer(self, input_utxos, fee_rate, sequence=BIP125_SEQUENCE_NUMBER):
+        """Given two input utxos, create one transaction that spends both of them"""
+        [tx, staging_tx] = list(map(lambda utxo: self.wallet.create_self_transfer(
```
As `input_utxos` is a pair instead of arbitrary length, does a function name such as `create_double_input_self_transfer` or `create_pair_input_self_transfer` indicate the intention better?

mjdietzx replied:
I think this is a good suggestion. I'll make this change as suggested

mjdietzx commented:
Rebased

mjdietzx committed:
test: incorrect rbf status when max replacement limit exceeded

Highlight that transactions may signal they are replaceable even
though they are not due to BIP125 RBF (Rule #5).

mjdietzx commented:
I better organized the commits and improved the commit message. Very minor / no behavior change

mjdietzx commented:
Rebased after #23437 and #22677 were merged

DariusParvin commented on test/functional/feature_rbf.py:
```diff
@@ -623,5 +642,130 @@ def test_replacement_relay_fee(self):
+                fee_rate=Decimal('0.0001'),
+            )
+            assert_equal(True, self.nodes[0].getmempoolentry(tx['txid'])['bip125-replaceable'])  # inherited
+        # Now we have a chain of: `optin_parent_tx`, `joined_tx`, and 99 txs. The last tx in the loop exceeded `MAX_REPLACEMENT_LIMIT`
```
nit: My understanding is that it's better not to have actual values in comments
```suggestion
        # Now we have a chain of: `optin_parent_tx`, `joined_tx`, and `MAX_REPLACEMENT_LIMIT - 1` txs. The last tx in the loop exceeded `MAX_REPLACEMENT_LIMIT`
```

DariusParvin commented on test/functional/feature_rbf.py:
```diff
@@ -623,5 +642,130 @@ def test_replacement_relay_fee(self):
+        assert_equal(True, self.nodes[0].getmempoolentry(tx["txid"])['bip125-replaceable'])
+
+        # Case in point, we can't actually replace `optin_parent_tx` once it has `MAX_REPLACEMENT_LIMIT` descendants
+        self.wallet.create_self_transfer(
```
nit: To check the specific error message
```suggestion
        replacement_parent_tx = self.wallet.create_self_transfer(
            from_node=self.nodes[0],
            utxo_to_spend=confirmed_utxos[0],
            fee_rate=Decimal('0.01'),
            mempool_valid=False
        )['hex']
        assert_raises_rpc_error(-26, 'too many potential replacements', self.nodes[0].sendrawtransaction, replacement_parent_tx, 0)
```

DariusParvin reviewed (commented):
Tested ACK 64c2040
Great job, these tests really clarify the current behavior of rbf.

darosior commented:
Concept ACK.

glozow reviewed (commented):
Thank you for making an effort to clarify the RBF interface, but I believe that clear documentation is more appropriate than a functional test for 2/4 of these.

aureleoules commented on test/functional/feature_rbf.py:
these asserts are redundant because of line 583 and 608
not sure if intended for readability

aureleoules reviewed (approved):
tACK 64c20402b258709af8c18eac1697939eca07082f (`make check` and `test/functional/test_runner.py`).
Tested on NixOS 22.05 64 bits.

## test_prechecks_overestimates_replacements

I commented the following lines locally and the test failed accordingly.  
https://github.com/bitcoin/bitcoin/blob/64c20402b258709af8c18eac1697939eca07082f/src/policy/rbf.cpp#L64-L69  


Increasing the following constant to 104 or more makes the test fail as expected because there would be enough room in the mempool to fit the chain of 104 transactions : `parent_txs (2) + joined_tx (2) + tx chain (100 / 2 but UTXOs counted twice so 100) = 104`.  
https://github.com/bitcoin/bitcoin/blob/64c20402b258709af8c18eac1697939eca07082f/src/policy/rbf.h#L17

## test_reorged_inherited_signaling_and_descendant_limit

Changing the following line
https://github.com/bitcoin/bitcoin/blob/64c20402b258709af8c18eac1697939eca07082f/src/policy/rbf.cpp#L34
to
```cpp
pool.CalculateMemPoolAncestors(entry, ancestors, MAX_BIP125_REPLACEMENT_CANDIDATES, noLimit, noLimit, noLimit, dummy, false);
```
results in the test failing as expected. The last transaction of the transaction chain of MAX_REPLACEMENT_LIMIT size would stop signaling RBF with inheritance because the child transaction signaling RBF would be buried too deep in the chain to be detected: which is to my understanding **not** the current behavior of Bitcoin.

kouloumos reviewed (commented):
ACK 5e72716a457e994c263dc1c226a4a1861cab6bbf, not supportive of the last 2 commits. 

I believe that 25f1e8aed1dea72ef2316ea5e39ca12d09c64921 has a lot of similarities with `test_opt_in`, it could be used to enchance that one.

kouloumos commented on test/functional/feature_rbf.py:
```diff
@@ -621,5 +637,64 @@ def test_replacement_relay_fee(self):
         tx.vout[0].nValue -= 1
         assert_raises_rpc_error(-26, "insufficient fee", self.nodes[0].sendrawtransaction, tx.serialize().hex())
 
+    def test_prechecks_overestimates_replacements(self):
```
What's documented here is not such an important issue as the CVE in `test_no_inherited_signaling` but documenting an issue/limitation seems to have a precedent. Maybe this could make sense as an additional reference for the docs?

glozow commented:
Closing as I believe this is resolved with doc/policy/mempool_replacements.md (which documents the limitations) and #25674, and this has needed rebase for a while.
